bokeh/events.py

```diff
@@ -7,90 +7,130 @@
+    may trigger at a very high rate.
 
 '''
 from __future__ import absolute_import
-from .util.future import with_metaclass
+
 import logging
 logger = logging.getLogger(__file__)
 
-EVENT_CLASSES = set()
-
+from .util.future import with_metaclass
 
+_CONCRETE_EVENT_CLASSES = dict()
 
-class MetaEvent(type):
+class _MetaEvent(type):
     ''' Metaclass used to keep track of all classes subclassed from Event.
 
-    All Event classes will be added to the EVENT_CLASSES set which is
-    used to decode event instances from JSON.
+    All Concrete Event classes (i.e. not "abstract" event base classes with
+    no ``event_name``) will be added to the _CONCRETE_EVENT_CLASSES set which
+    is used to decode event instances from JSON.
+
     '''
     def __new__(cls, clsname, bases, attrs):
-        newclass = super(MetaEvent, cls).__new__(cls, clsname, bases, attrs)
-        EVENT_CLASSES.add(newclass)
+        newclass = super(_MetaEvent, cls).__new__(cls, clsname, bases, attrs)
+        if newclass.event_name is not None:
+            _CONCRETE_EVENT_CLASSES[newclass.event_name] = newclass
         return newclass
 
-
-class Event(with_metaclass(MetaEvent, object)):
+class Event(with_metaclass(_MetaEvent, object)):
     ''' Base class for all Bokeh events.
 
+    This base class is not typically useful to instantiate on its own.
+
     '''
     _event_classes = []
     event_name = None
 
     def __init__(self, model):
+        ''' Create a new base event.
+
+        Args:
+
+            model (Model) : a Bokeh model to register event callbacks on
+
+        '''
         self._model_id = None
         if model is not None:
             self._model_id = model._id
 
     @classmethod
     def decode_json(cls, dct):
-        ''' Custom json decoder for Events for use with the object_hook
-        argument of json.load or json.loads.
+        ''' Custom JSON decoder for Events.
+
+        Can be used as the ``object_hook`` argument of ``json.load`` or
+        ``json.loads``.
+
+        Args:
+            dct (dict) : a JSON dictionary to decode
+                The dictionary should have keys ``event_name`` and ``event_values``
+
+        Raises:
+            ValueError, if the event_name is unknown
+
+        Examples:
+
+            .. code-block:: python
+
+                >>> import json
+                >>> from bokeh.events import Event
+                >>> data = '{"event_name": "pan", "event_values" : {"model_id": 1, "x": 10, "y": 20, "sx": 200, "sy": 37}}'
+                >>> json.loads(data, object_hook=Event.decode_json)
+                <bokeh.events.Pan object at 0x1040f84a8>
+
         '''
         if not (('event_name' in dct) and ('event_values' in dct)):
             return dct
 
+        event_name = dct['event_name']
+
+        if event_name not in _CONCRETE_EVENT_CLASSES:
+            raise ValueError("Could not find appropriate Event class for event_name: %r" % event_name)
```

Comment:
This is the change. If the class can't be found, an exception is now raised. Since this should only be used in the context of our low level code decoding protocol messages, this should really never happen. I'd prefer if it *does* happen for it to fail bigly.